src/compute/src/render/continual_task.rs

```diff
@@ -107,18 +111,119 @@ pub(crate) struct ContinualTaskCtx<G: Scope<Timestamp = Timestamp>> {
     name: Option<String>,
     dataflow_as_of: Option<Antichain<Timestamp>>,
     ct_inputs: BTreeSet<GlobalId>,
+    ct_outputs: BTreeSet<GlobalId>,
     pub ct_times: Vec<Collection<G, (), Diff>>,
 }
 
+pub(crate) enum ContinualTaskSourceTransformer {
+    InsertsInput,
+    SelfReference { source_id: GlobalId },
+    NormalReference,
+}
+
+impl ContinualTaskSourceTransformer {
+    pub fn snapshot_mode(&self) -> SnapshotMode {
+        use ContinualTaskSourceTransformer::*;
+        match self {
+            InsertsInput => SnapshotMode::Exclude,
+            SelfReference { .. } | NormalReference { .. } => SnapshotMode::Include,
+        }
+    }
+
+    pub fn transform<S: Scope<Timestamp = Timestamp>>(
+        &self,
+        oks: Collection<S, Row, Diff>,
+        errs: Collection<S, DataflowError, Diff>,
+    ) -> (
+        Collection<S, Row, Diff>,
+        Collection<S, DataflowError, Diff>,
+        Collection<S, (), Diff>,
+    ) {
+        use ContinualTaskSourceTransformer::*;
+        match self {
+            // Make a collection s.t, for each time T in the input, the output
+            // contains the inserts at T.
+            InsertsInput => {
+                // Keep only the inserts.
+                //
+                // At some point this will become a user option to instead keep
+                // only deletes.
+                let oks = oks.inner.filter(|(_, _, diff)| *diff > 0);
+                // Grab the original times for use in the sink operator.
+                let times = oks.map(|(_row, ts, diff)| ((), ts, diff));
```

Comment:
This fork in the dataflow graph double memory requirements. Ideally, we have an operator that has two outputs, or we work on shared references (`stream.shared()`). This can be problematic mostly for an initial snapshot.

Reply:
Sounds good, thanks!

Reply:
Done! https://github.com/MaterializeInc/materialize/pull/30123